wit/types.wit

```diff
@@ -1,41 +1,44 @@
-/// An interface grouping all necessary types for messaging.
 default interface messaging-types {
-    /// A broker type that allows the exchange of messages.
-    type broker = u32
-    drop-broker: func(b: broker)
-    open-broker: func(name: string) -> result<broker, error>
-
-    /// An event type that follows the CloudEvents specification (https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md). We
-    /// assume the type of the data is a byte sequence. It is up to the data schema to determine what type of the data payload the event 
-    /// contains.
-    record event {
-        specversion: string,
-        ty: string,
-        source: string,
-        id: string,
-        data: option<list<u8>>,
-        datacontenttype: option<string>,
-        dataschema: option<string>,
-        subject: option<string>,
-        time: option<string>,	
+    /// A connection to a message-exchange service (e.g., buffer, broker, etc.).
+    type client = u32
+    disconnect: func(c: client)
+    connect: func(name: string) -> result<client, error>
+  
+    /// TODO(danbugs): This should be eventually extracted as an underlying type for other wasi-cloud-core interfaces.
+    type error = u32
+    drop-error: func(e: error)
+    trace: func(e: error) -> string
+  
+    /// There are two types of channels:
+    /// - publish-subscribe channel, which is a broadcast channel, and
+    /// - point-to-point channel, which is a unicast channel.
+    ///
+    /// The interface doesn't highlight this difference in the type itself as that's uniquely a consumer issue.
+    type channel = string
+  
+    /// Configuration includes a required list of channels the guest is subscribing to, and an optional list of extensions key-value pairs 
+    /// (e.g., partitions/offsets to read from in Kafka/EventHubs, QoS etc.).
+    record guest-configuration {
+        channels: list<channel>,
         extensions: option<list<tuple<string, string>>>
     }
-    
-    /// Channels specify where a published message should land. There are two types of channels:
-    /// - queue: competitive consumers, and
-    /// - topic: non-competitive consumers.
-    variant channel {
-        queue(string),
-        topic(string)
+  
+    /// Format specification for messages 
+    ///  - more info: https://github.com/clemensv/spec/blob/registry-extensions/registry/spec.md#message-formats
+    ///  - message metadata can further decorate w/ things like format version, and so on.
+    enum format-spec {
```

Comment:
As mentioned in issue #8 , I don't think the payload type should be surfaced as part of the interface. If we do this, then we'll have to change the interface every time we find a new payload type that we want to support or, more likely, everyone will just use `raw`.

I don't expect my brokers to care what shape my bytes take internally, I only expect them to deliver my bytes according to my needs.

Reply:
Another use case that this format spec can interfere with is using different transports. Let's say I have a component and I want to run it both in a cloud and on a constrained edge device. In the cloud, this component can use a robust messaging transport. On the constrained device, it probably uses MQTT. The component should not need to know if it's running in the cloud or a constrained device, it should simply execute its business logic while the use of MQTT or Kafka or anything else is an invisible detail. If we have to specifically embed the format type into our compiled code, it blocks the benefits of portability that WebAssembly is so famous for.

Reply:
What are your thoughts on simply making `format-spec` a `string` alias then instead of an enum?~ similar to `content-type` in HTTP headers.

Reply:
Why don't we just call it `content-type` and make it a string, this way it's easier for people to make that analogy?

Reply:
For an API abstraction, you will want to have access to the features of the broker you want to deal with. For a common event format like CloudEvents, you want to abstract as much of that away. If you want a common envelope for use with an API, pick a common envelope format like CNCF CloudEvents. We* did a lot of work to find commonalities and figure out transport bindings and mappings. 

A unified API that works across products should have a foundational message that is even more fundamental than CloudEvents (payload content-type, payload body, metadata map and map of maps) so that anyone can understand the payload without having to understand envelopes or protocols at all. For most apps, payload is the most important thing.

Yet, the unified API should not aim to hide the details of the underlying brokers and their specific metadata because if it does, developers will quickly jump back to product-specific APIs. The experience needs to provide a balance of unified experience and maximum product/feature access fidelity at the same time. If you use Kafka, you need to be able to set the record/partition key and if you use an AMQP broker you need to be able to set the group-id. 

Therefore, having a selector that tells you want to expect in the metadata section, similar to content-type helping with what to expect in the payload section, does make sense. 

* (I am one of the architects of CNCF CloudEvents)